plugins/metrics/pkg/agent/node.go

```diff
@@ -149,13 +171,66 @@ func (m *MetricsNode) GetHealth(_ context.Context, _ *emptypb.Empty) (*corev1.He
 	}, nil
 }
 
+// Start Implements remoteread.RemoteReadServer
+
+func (m *MetricsNode) Start(ctx context.Context, request *remoteread.StartReadRequest) (*emptypb.Empty, error) {
+	m.targetRunnerMu.Lock()
+	defer m.targetRunnerMu.Unlock()
+
+	if err := m.targetRunner.Start(request.Target, request.Query); err != nil {
+		return nil, err
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func (m *MetricsNode) Stop(ctx context.Context, request *remoteread.StopReadRequest) (*emptypb.Empty, error) {
+	m.targetRunnerMu.Lock()
+	defer m.targetRunnerMu.Unlock()
+
+	if err := m.targetRunner.Stop(request.Meta.Name); err != nil {
+		return nil, err
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+func (m *MetricsNode) GetTargetStatus(ctx context.Context, request *remoteread.TargetStatusRequest) (*remoteread.TargetStatus, error) {
+	m.targetRunnerMu.RLock()
+	defer m.targetRunnerMu.RUnlock()
+
+	return m.targetRunner.GetStatus(request.Meta.Name)
+}
+
+func (m *MetricsNode) Discover(ctx context.Context, request *remoteread.DiscoveryRequest) (*remoteread.DiscoveryResponse, error) {
+	driver, err := drivers.GetNodeDriver("external-operator")
+	if err != nil {
+		return nil, fmt.Errorf("could not get driver: %w", err)
+	}
+
+	promDriver, ok := driver.(*drivers.ExternalPromOperatorDriver)
+	if !ok {
+		m.logger.Errorf("found wrong driver type '%T'", driver)
+		return nil, fmt.Errorf("could not discover Prometheus instances")
+	}
+
+	entries, err := promDriver.Discover(*request.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("could not discover Prometheus instances: %w", err)
+	}
+
+	return &remoteread.DiscoveryResponse{
+		Entries: entries,
+	}, nil
+}
+
 func (m *MetricsNode) doSync(ctx context.Context) {
 	m.logger.Debug("syncing metrics node")
 	m.nodeClientMu.RLock()
 	defer m.nodeClientMu.RUnlock()
 
 	if m.nodeClient == nil {
-		m.conditions.Set(health.CondConfigSync, health.StatusPending, "no client, skipping sync")
+		m.conditions.Set(health.CondConfigSync, health.StatusPending, "no promClient, skipping sync")
```

Comment:
Is it possible to get here? Looks like doSync is only called if nodeClient is not nil in the first place.

Reply:
I think the only change I made here was accidentally chaning the `client` -> `promClient`.

Regardless I think in [MetricsNode.SetNodeClient](https://github.com/rancher/opni/blob/ff5c37f4522a74ad086bf86208f53fd876d78238/plugins/metrics/pkg/agent/node.go#L95-L102) its possible to pass a `nil` client. The only place its called is with a non-nil client, but its *possible*. It should be safe to remove the check, but I'd rather leave it in or move the check into `SetNodeClient`.